### DISCUSSION TITLE
Multi instance container setup and repo access

### DISCUSSION THREAD

**Discussion by codeplay87** [oracle/opengrok#4334] [General]
I am trying to understand how opengrok mirroring and indexing works to validate that it won't cause any issues if we have multiple docker instances pointing at shared volume. 
 
We currently have the following setup in k8s on azure:
- Persistent Volume for the opengrok (src, data, etc) 
- Persistent Volume for the tomcat server config.xml

If we had n* of instances running against the same shared data does opengrok do any locking of those repos when it runs its mirroring or indexing?

Many thanks in advance.

**Comment by vladak** [comment]
The OpenGrok tooling (`opengrok-mirror` and `opengrok-sync`) use file based locking in system temporary directory (usually `/tmp` in Linux distros). If projects are enabled, the lock files will be specific to given project.

As for the indexer itself, one can set the Lucene locking style with the `--lock` option. The default is off, i.e. no locking. The other choices are either `simple` or `native` which corresponds to Lucene's [SimpleFSLockFactory](https://lucene.apache.org/core/8_1_0/core/org/apache/lucene/store/SimpleFSLockFactory.html) and [NativeFSLockFactory](https://lucene.apache.org/core/8_1_0/core/org/apache/lucene/store/NativeFSLockFactory.html), respectively. I believe the former will use lock files, while the latter will employ `fcntl()` style locking (on Unix-like systems).

**Reply by codeplay87** [reply]
Thanks for getting back so quickly. Is there anyway to set the location of that directory used for the file based locking for mirror and sync?

**Reply by vladak** [reply]
Not at the moment, e.g. for `opengrok-mirror` it explicitly uses `tempfile.gettempdir()`: https://github.com/oracle/opengrok/blob/f075707db31c7e8efcd23f075e4f6611671aace9/tools/src/main/python/opengrok_tools/mirror.py#L212-L214

Under the covers, `filelock` will use whatever implementation it can find, preferring native (alluding to the Lucene locking), i.e. `fcntl()` based locking on Unix-like systems or if it is not found, it will fall back to lock files (called "soft locking" there): https://github.com/tox-dev/py-filelock/blob/dbe3ba1470838f4aded968c8d364d6e8ab6c2c00/src/filelock/__init__.py#L27-L31
